volatility3/framework/symbols/windows/extensions/consoles.py

```diff
@@ -0,0 +1,439 @@
+
+        The screen buffer is comprised of the screen's Y
+        coordinate which tells us the number of rows and
+        the X coordinate which tells us the width of each
+        row in characters. Windows 10 17763 changed from
+        a large text buffer to a grid of cells, with each
+        cell containing a single wide character in that
+        cell, stored in a CharRowCell object.
+
+        @param truncate: True if the empty rows at the
+        end (i.e. bottom) of the screen buffer should be
+        supressed.
+        """
+        rows = []
+
+        capacity = self.TextBufferInfo.BufferCapacity
+        start = self.TextBufferInfo.BufferStart
+        buffer_rows = self.TextBufferInfo.BufferRows.dereference()
+        buffer_rows.Rows.count = self.TextBufferInfo.BufferCapacity
+
+        for i in range(capacity):
+            index = (start + i) % capacity
+            row = buffer_rows.Rows[index]
+            if hasattr(row, "Row"):
+                row = row.Row
+            try:
+                text = row.get_text(truncate_lines)
+                rows.append(text)
+            except:
+                break
+
+        if truncate_rows:
+            rows = self._truncate_rows(rows)
+
+        if rows:
+            rows = ["=== START OF BUFFER ==="] + rows + ["=== END OF BUFFER ==="]
+        else:
+            rows = ["=== NO BUFFER DATA FOUND  ==="]
+        return rows
+
+
+class CONSOLE_INFORMATION(objects.StructType):
+    """A Console Information Structure."""
+
+    @property
+    def ScreenBuffer(self):
+        return self.GetScreenBuffer
+
+    def is_valid(self, max_buffers=4) -> bool:
+        """Determine if the structure is valid."""
+
+        # Last displayed must be between -1 and max
+        if self.HistoryBufferCount < 1 or self.HistoryBufferCount > max_buffers:
+            return False
+
+        if not self.get_title() and not self.get_original_title():
+            return False
+
+        return True
+
+    def get_screens(self):
+        """Generator for screens in the console.
+
+        A console can have multiple screen buffers at a time,
+        but only the current/active one is displayed.
+
+        Multiple screens are tracked using the singly-linked
+        list _SCREEN_INFORMATION.Next.
+
+        See CreateConsoleScreenBuffer
+        """
+        screens = [self.CurrentScreenBuffer]
+
+        if self.ScreenBuffer not in screens:
+            screens.append(self.ScreenBuffer)
+
+        seen = set()
+
+        for screen in screens:
+            cur = screen
+            while cur and cur.vol.offset != 0 and cur.vol.offset not in seen:
+                cur.TextBufferInfo.BufferRows.Rows.count = (
+                    cur.TextBufferInfo.BufferCapacity
+                )
+                yield cur
+                seen.add(cur.vol.offset)
+                cur = cur.Next
+
+    def get_histories(self):
+        for cmd_hist in self.HistoryList.dereference().to_list(
```

Comment:
Again, this should dereference properly unless the Pointer class has a `to_list` method of its own.

Reply:
also removed